addons/binding/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/handler/EnergenieGatewayHandler.java

```diff
@@ -0,0 +1,619 @@
+                    TimeUnit.SECONDS);
+        } catch (RejectedExecutionException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    REJECTED_EXECUTION_OF_UPDATE_TASK_MESSAGE);
+        }
+    }
+
+    /**
+     * Checks when the gateway was last active and updates the thing's channel.
+     * This method is called periodically by a refresh thread.
+     */
+    public void checkGateway() {
+        /*
+         * Mi|Home REST API does not provide a method for showing a gateway's
+         * information. So in order to check when was the last time it was seen
+         * you need to list all gateways and find the one you want by its id.
+         */
+        JsonObject gateway = getGatewayById(gatewayId);
+        if (gateway != null) {
+            setProperties(gateway);
+            gatewayLabel = gateway.get(DeviceConstants.DEVICE_LABEL_KEY).getAsString();
+            thing.setLabel(gatewayLabel);
+
+            JsonElement lastSeenElement = gateway.get(DeviceConstants.GATEWAY_LAST_SEEN_KEY);
+            if (!lastSeenElement.isJsonNull()) {
+                String lastSeenString = lastSeenElement.getAsString();
+                handleCommand(thing.getChannel(EnergenieBindingConstants.CHANNEL_LAST_SEEN).getUID(),
+                        DateTimeType.valueOf(lastSeenString));
+
+                handleLastSeenProperty(lastSeenString);
+            }
+        }
+    }
+
+    /**
+     * Searches in the list of all registered gateways and find the one with the
+     * given id
+     *
+     * @param id
+     *            - the id of the searched gateway
+     * @return the gateway with the given id if it is registered or null if it
+     *         is not
+     */
+    private JsonObject getGatewayById(int searchedId) {
+        JsonObject response = apiManager.listGateways();
+        if (response != null) {
+            JsonArray allGateways = response.getAsJsonArray(JSONResponseConstants.DATA_KEY);
+            if (allGateways != null) {
+                for (JsonElement gateway : allGateways) {
+                    JsonElement id = gateway.getAsJsonObject().get(DeviceConstants.DEVICE_ID_KEY);
+                    if (!id.isJsonNull() && id.getAsInt() == searchedId) {
+                        return gateway.getAsJsonObject();
+                    }
+                }
+                logger.warn(UNREGISTERED_GATEWAY_MESSAGE);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, UNREGISTERED_GATEWAY_MESSAGE);
+                setGatewayAlreadyUnregistered(true);
+                return null;
+            }
+        }
+        return null;
+    }
+
+    private void handleLastSeenProperty(String lastSeenString) {
+        Long timeInterval = null;
+        String currentDateTimeString = new SimpleDateFormat(EnergenieBindingConstants.DATE_TIME_PATTERN)
```

Comment:
For another time, another binding, Java 8 introduced a proper thread safe formatter: `DateTimeFormatter`